Run risky code in finally block

greg0ire opened the pull request:
`catch` blocks are not supposed to fail. If you want to do something despite an exception happening, you should do it in a `finally` block.

(finally) Fixes #7545

greg0ire committed:
Run risky code in finally block

catch blocks are not supposed to fail. If you want to do something
despite an exception happening, you should do it in a finally block.

Closes #7545

greg0ire commented:
Should I move the call to `commit()` inside the `finally` block as well? :thinking:

SenseException reviewed (dismissed):
> Should I move the call to commit() inside the finally block as well? 🤔

I don't know. It looks like it would all be the same in your changes. Logically, from a reading perspective, it would look weird if the commit would be in `finally`

Does anyone know why `$conn->isTransactionActive()` wasn't used in the EntityManager in the past?

ro0NL commented:
as a transaction was started, it should be active during rollback

if not it exposes userland closes more than transaction than it opened

greg0ire commented:
I talked a bit more with @alcaeus and it wouldn't be problematic from a reading perspective only: if an exception happens after the actual commit, you don't want to rollback. I'll try adding the guard clause as well as tests.

greg0ire committed:
Add guard clause

It maybe happen that the SQL COMMIT statement is successful, but then
something goes wrong. In that kind of case, you do not want to attempt a
rollback.

This was implemented in UnitOfWork::commit(), but for some reason not in
the similar EntityManager methods.
